src/v/pandaproxy/schema_registry/seq_writer.cc

```diff
@@ -70,37 +139,36 @@ ss::future<> seq_writer::wait_for(model::offset offset) {
 /// Helper for write methods that need to check + retry if their
 /// write landed where they expected it to.
 ///
-/// \param write_at Offset at which caller expects their write to land
+/// \param write_at Offset at which caller expects their write to land. If
+/// std::nullopt, the offset is not checked.
 /// \param batch Message to write
 /// \return true if the write landed at `write_at`, else false
-ss::future<bool> seq_writer::produce_and_check(
-  model::offset write_at, model::record_batch batch) {
-    // Because we rely on checking exactly where our message (singular) landed,
-    // only use this function with batches of a single message.
-    vassert(batch.record_count() == 1, "Only single-message batches allowed");
+ss::future<bool> seq_writer::produce_and_apply(
+  std::optional<model::offset> write_at, model::record_batch batch) {
+    vassert(
+      write_at.value_or(batch.base_offset()) == batch.base_offset(),
+      "Set the base_offset to the expected write_at");
```

Comment:
qq: Rather than the optional, why not some boolean flag that indicates if you want the written offset checked or not?  If `write_at` has to be equal to `batch.base_offset()` otherwise it runtime asserts, wouldn't that introduce the possibility of instability?

Reply:
> qq: Rather than the optional, why not some boolean flag that indicates if you want the written offset checked or not? If `write_at` has to be equal to `batch.base_offset()` otherwise it runtime asserts, wouldn't that introduce the possibility of instability?

The main change in that area is that multiple records can go in a batch and get written in one go.

Currently, if there are multiple records, only the first one will require the offset to be checked, if any other checked offset record is added to the batch, it will fail.

I did play around with attempting to automagically adding the checked offset in the batch_builder, but it was hard to reason about. The API here isn't great, but I figured any problem would be found pretty quickly in tests.